sway-lsp/src/core/session.rs

```diff
@@ -60,6 +61,11 @@ pub struct Session {
     pub type_engine: RwLock<TypeEngine>,
     pub decl_engine: RwLock<DeclEngine>,
     pub sync: SyncWorkspace,
+    // Limit the number of threads that can wait to parse at the same time. One thread can be parsing
+    // and one thread can be waiting to start parsing. All others will return the cached diagnostics.
+    parse_permits: Arc<Semaphore>,
+    // Cached diagnostic results that require a lock to access. Readers will wait for writers to complete.
+    diagnostics: Arc<RwLock<Diagnostics>>,
```

Comment:
As joshy mentioned in the sync, it'd be nice if tower-lsp had a sync option (rather than assuming an async API) so we wouldn't have to manually sync access like this! Alas, this looks fine in the meantime :+1:

Reply:
It turns out there is kind of a sync option: we can set max_concurrency=1. When I tried it out locally, the LSP ran much slower. I also read that `cancelRequest` only works when tower-lsp is running with multiple threads. 

I think we'd need to bring compilation time down a lot for the LSP to be usable running synchronously. From what I can tell, most of the problems with async LSP stem from the ~1 second it takes to compile the project, since LSP needs to constantly recompile when anything changes in the editor.